### PR TITLE
[MTDSA-9086] - Dependency upgrades including play 28

### PR DESCRIPTION
**Pull request by DreamcastKnight**

### CODE REVIEW

File: app/config/AppConfig.scala [hunk 1]
```diff
@@ -28,10 +28,12 @@ trait AppConfig {
   def desBaseUrl: String
   def desEnv: String
   def desToken: String
+  def desEnvironmentHeaders: Option[Seq[String]]
 
   def ifsBaseUrl: String
   def ifsEnv: String
   def ifsToken: String
+  def ifsEnvironmentHeaders: Option[Seq[String]]
```
Comment (Mbrookesy) [inline app/config/AppConfig.scala:36]: Could you add comments above both the DES and IFS lines on line 28 and 33 saying something for each like "DES/IFS Config"

Comment (Mbrookesy) [reply app/config/AppConfig.scala:36]: Could you also add a comment for "API Config" above the API config variables on line 41

File: app/config/AppConfig.scala [hunk 2]
```diff
@@ -51,10 +53,12 @@ class AppConfigImpl @Inject()(config: ServicesConfig, configuration: Configurati
   val desBaseUrl: String = config.baseUrl("des")
   val desEnv: String = config.getString("microservice.services.des.env")
   val desToken: String = config.getString("microservice.services.des.token")
+  val desEnvironmentHeaders: Option[Seq[String]] = configuration.getOptional[Seq[String]]("microservice.services.des.environmentHeaders")
 
   val ifsBaseUrl: String = config.baseUrl("ifs")
   val ifsEnv: String = config.getString("microservice.services.ifs.env")
   val ifsToken: String = config.getString("microservice.services.ifs.token")
+  val ifsEnvironmentHeaders: Option[Seq[String]] = configuration.getOptional[Seq[String]]("microservice.services.ifs.environmentHeaders")
```
Comment (Mbrookesy) [inline app/config/AppConfig.scala:61]: Could you add similar comments to these ones as well matching up with the DES, IFS and API lines

File: build.sbt [hunk 3]
```diff
@@ -44,12 +44,12 @@ lazy val microservice = Project(appName, file("."))
...
-    parallelExecution in ItTest := false,
+    ItTest / fork := true,
+    ItTest / unmanagedSourceDirectories := Seq((baseDirectory in ItTest).value / "it"),
+    ItTest / unmanagedClasspath += baseDirectory.value / "resources",
+    Runtime / unmanagedClasspath += baseDirectory.value / "resources",
+    ItTest / javaOptions += "-Dlogger.resource=logback-test.xml",
+    ItTest / parallelExecution := false,
     addTestReportOption(ItTest, "int-test-reports")
   )
   .settings(
```
Comment (Mbrookesy) [inline build.sbt:55]: Line 64 can be deleted as domain is now deprecated

File: build.sbt [hunk 4]
```diff
@@ -44,12 +44,12 @@ lazy val microservice = Project(appName, file("."))
...
     inConfig(ItTest)(Defaults.itSettings ++ headerSettings(ItTest) ++ automateHeaderSettings(ItTest)),
-    fork in ItTest := true,
-    unmanagedSourceDirectories in ItTest := Seq((baseDirectory in ItTest).value / "it"),
-    unmanagedClasspath in ItTest += baseDirectory.value / "resources",
-    unmanagedClasspath in Runtime += baseDirectory.value / "resources",
-    javaOptions in ItTest += "-Dlogger.resource=logback-test.xml",
-    parallelExecution in ItTest := false,
+    ItTest / fork := true,
+    ItTest / unmanagedSourceDirectories := Seq((baseDirectory in ItTest).value / "it"),
```
Comment (Mbrookesy) [inline build.sbt:48]: `baseDirectory in ItTest` needs to be updated to `ItTest / baseDirectory`

File: it/v1/endpoints/AddCustomEmploymentControllerISpec.scala [hunk 5]
```diff
@@ -260,8 +260,8 @@ class AddCustomEmploymentControllerISpec extends IntegrationBaseSpec {
 
       val missingMandatoryFieldErrors: MtdError = RuleIncorrectOrEmptyBodyError.copy(
         paths = Some(List(
-          "/startDate",
-          "/employerName"
+          "/employerName",
+          "/startDate"
```
Comment (Mbrookesy) [inline it/v1/endpoints/AddCustomEmploymentControllerISpec.scala:264]: Could you explain for the controller and validation parts of the pr, why these were changed?

It seems very out of scope these changes.

File: project/AppDependencies.scala [hunk 6]
```diff
@@ -22,20 +22,20 @@ object AppDependencies {
 
   val compile: Seq[ModuleID] = Seq(
     ws,
-    "uk.gov.hmrc"        %% "bootstrap-backend-play-26" % "3.4.0",
+    "uk.gov.hmrc"        %% "bootstrap-backend-play-28" % "4.3.0",
```
Comment (Mbrookesy) [inline project/AppDependencies.scala:25]: play 28 can be updated to 5.3.0

File: project/AppDependencies.scala [hunk 7]
```diff
@@ -22,20 +22,20 @@ object AppDependencies {
...
     "uk.gov.hmrc"        %% "domain"                    % "5.11.0-play-26",
-    "org.typelevel"      %% "cats-core"                 % "2.5.0",
+    "org.typelevel"      %% "cats-core"                 % "2.6.0",
```
Comment (Mbrookesy) [inline project/AppDependencies.scala:27]: cats-core can be updated to 2.6.1

File: project/AppDependencies.scala [hunk 8]
```diff
@@ -22,20 +22,20 @@ object AppDependencies {
 
   val compile: Seq[ModuleID] = Seq(
     ws,
-    "uk.gov.hmrc"        %% "bootstrap-backend-play-26" % "3.4.0",
+    "uk.gov.hmrc"        %% "bootstrap-backend-play-28" % "4.3.0",
     "uk.gov.hmrc"        %% "domain"                    % "5.11.0-play-26",
```
Comment (Mbrookesy) [inline project/AppDependencies.scala:26]: domain can be deleted as it's deprecated

File: project/AppDependencies.scala [hunk 9]
```diff
@@ -22,20 +22,20 @@ object AppDependencies {
...
-    "org.typelevel"      %% "cats-core"                 % "2.5.0",
+    "org.typelevel"      %% "cats-core"                 % "2.6.0",
     "com.chuusai"        %% "shapeless"                 % "2.4.0-M1",
     "com.neovisionaries" % "nv-i18n"                    % "1.27"
```
Comment (Mbrookesy) [inline project/AppDependencies.scala:29]: This can be removed.

File: project/AppDependencies.scala [hunk 10]
```diff
@@ -22,20 +22,20 @@ object AppDependencies {
...
   )
 
   def test(scope: String = "test, it"): Seq[sbt.ModuleID] = Seq(
-    "org.scalatest"          %% "scalatest"          % "3.2.3"             % scope,
+    "org.scalatest"          %% "scalatest"          % "3.2.7"             % scope,
```
Comment (Mbrookesy) [inline project/AppDependencies.scala:33]: This can be updated to 3.2.9

File: project/AppDependencies.scala [hunk 11]
```diff
@@ -22,20 +22,20 @@ object AppDependencies {
...
     "com.vladsch.flexmark"   % "flexmark-all"        % "0.36.8"            % scope,
     "org.scalacheck"         %% "scalacheck"         % "1.15.1"            % scope,
```
Comment (Mbrookesy) [inline project/AppDependencies.scala:35]: scalacheck can be 1.15.4

File: test/v1/connectors/AddCustomEmploymentConnectorSpec.scala [hunk 12]
```diff
@@ -66,8 +67,10 @@ class AddCustomEmploymentConnectorSpec extends ConnectorSpec {
         MockedHttpClient
           .post(
             url = s"$baseUrl/income-tax/income/employments/$nino/$taxYear/custom",
+            config = dummyDesHeaderCarrierConfig,
             body = addCustomEmploymentRequestBody,
-            requiredHeaders = requiredDesHeaders :_*
+            requiredHeaders = requiredDesHeaders,
+            excludedHeaders = Seq("AnotherHeader" -> "HeaderValue")
```
Comment (Mbrookesy) [inline test/v1/connectors/AddCustomEmploymentConnectorSpec.scala:73]: [Applies to all the connector specs]

For the connector specs that have a body (PUT/POST) you should add a line before the MockedHttpClient line (For example in this one that would be Line 67)

Something like this: `val requiredDesHeadersPost: Seq[(String, String)] = requiredDesHeaders ++ Seq("Content-Type" -> "application/json")` which basically adds the Content-Type header to it as that would be included in the live version.

Comment (Mbrookesy) [reply test/v1/connectors/AddCustomEmploymentConnectorSpec.scala:73]: Could you also include something like above the MockedHttpClient for all of them `implicit val hc: HeaderCarrier = HeaderCarrier(otherHeaders = otherHeaders ++ Seq("Content-Type" -> "application/json"))`

File: test/v1/connectors/ConnectorSpec.scala [hunk 13]
```diff
@@ -27,12 +27,44 @@ trait ConnectorSpec extends UnitSpec
...
 
   implicit val correlationId: String = "a1e8057e-fbbc-47a8-a8b4-78d9f015c253"
   implicit val hc: HeaderCarrier = HeaderCarrier()
   implicit val ec: ExecutionContext = scala.concurrent.ExecutionContext.global
 
+  val dummyDesHeaderCarrierConfig: HeaderCarrier.Config =
+    HeaderCarrier.Config(
+      Seq("^not-test-BaseUrl?$".r),
+      Seq.empty[String],
+      Some("self-assessment-accounts-api")
```
Comment (Mbrookesy) [inline test/v1/connectors/ConnectorSpec.scala:40]: This should be `individuals-income-received-api`

File: test/v1/connectors/ConnectorSpec.scala [hunk 14]
```diff
@@ -27,12 +27,44 @@ trait ConnectorSpec extends UnitSpec
...
+    )
+
+  val dummyIfsHeaderCarrierConfig: HeaderCarrier.Config =
+    HeaderCarrier.Config(
+      Seq("^not-test-BaseUrl?$".r),
+      Seq.empty[String],
+      Some("self-assessment-accounts-api")
```
Comment (Mbrookesy) [inline test/v1/connectors/ConnectorSpec.scala:47]: This should be `individuals-income-received-api`

**Review by Mbrookesy — CHANGES_REQUESTED** [review]
Currently the ATs are failing against the branch, seems to be a lot of rule incorrect body errors.

File: it/v1/endpoints/AddCustomEmploymentControllerISpec.scala (hunk 5 above)
Comment (DreamcastKnight) [reply it/v1/endpoints/AddCustomEmploymentControllerISpec.scala:264]: for some reason, the tests started failing after my changes, switching around the order all the errors were. I hadn't any idea why this was happening, so I changed the tests.

**Pull request merged**